### PR TITLE
[Fix #916] New Templating system

### LINKED ISSUE #916: Convert templates to real models
Since templates are now code in AS, we can make them code in the service as well. This will allow us to use django's form validation for #655 , use foreign key relationships for the image gallery (#841) and property fix things like #757 

Let's investigate how that would look like and that's the best route to follow.

## Check 
 - [x] Save as new functionality
 - [x] Duplicate
 - [x] user permissions
 - [x] New code serves old templates if no `template_relation` exists

## Deployment plan
 - [x] Backup production DB
 - [x] Run the [migration script](https://gist.github.com/glogiotatidis/f143603e80e9044e0a9e67d777356147) locally with the backup of the previous step. Confirm that all fields are transferred correctly, especially images.
  - List of snippets are required manual migration (all in draft state):
    - [x] 11395
    - [x] 11421
    - [x] 11424
### Staging
 - [x] Set `MEDIA_ICONS_ROOT` config stage
 - [x] Push code to staging
 - [x] Update staging user permissions
 - [x] kubectl into the staging node and install inkscape, required for the migration
   - [x] k8s_oregon_b -n snippets-stage plugin exec-user snippets-stage-clock-86cf445477-7wrjb bash
 - [x] Run migration script
 - [x] Clear bundle cache
   - [x] `deis_oregon run -a snippets-stage  -- ./manage.py clear_cache -a`
 - [x] Confirm bundle generation
 - [x] Kill + restart migration pod
   - [x] k8s_oregon_b -n snippets-stage delete po/snippets-stage-clock-86cf445477-7wrj
### Production
 - [x] Set `MEDIA_ICONS_ROOT` config prod 
   - [x] admin
   - [x] prod oregon
   - [x] prod frankfurt
 - [x] Push code to prod
 - [x] Update prod user permissions
   - [x] Edit ASR Snippets group, search for templates, add all 
   - [x] Edit ASR Snippets group, search for icon, add all
 - [x] kubectl into the staging node and install inkscape, required for the migration
 - [x] Run migration script
 - [x] Clear bundle cache
   - [x] `deis_oregon run -a snippets-admin  -- ./manage.py clear_cache -a`
   - [x] `deis_frankfurt run -a snippets-prod  -- ./manage.py clear_cache -a`
   - [x] `deis_oregon run -a snippets-prod  -- ./manage.py clear_cache -a`
 - [x] Confirm bundle generation
 - [x] Kill + restart migration pod

### PR DESCRIPTION
**Pull request by glogiotatidis**
Note to self:
 - [x] Check that default values like 'block button text' are set.
 - [x] Check that bundle keys update when images update

### CODE REVIEW

File: snippets/base/admin/adminmodels.py [hunk 1]
```diff
@@ -103,6 +103,81 @@ class AddonAdmin(admin.ModelAdmin):
...
+        fields = super().get_readonly_fields(request, obj)
+
+        # Don't allow image to be changed after set.
+        if obj and 'image' not in fields:
+            fields.append('image')
+
+        return fields
+
+
+class SimpleTemplateInline(admin.StackedInline):
```
Comment (glogiotatidis) [inline snippets/base/admin/adminmodels.py]: Not used yet, will add this inline along with some UX sauce in a following PR

**Comment by metadave** [comment]
> ... I'm pretty lost as to what's going on and why we need new templates and whatnot

seems like a good candidate for an ADR so we can capture why this is happening

**Comment by pmac** [comment]
I guess so. It's not exactly an architectural change though IMHO. But I would appreciate a bug or issue at least detailing what's happening and why.

**Comment by metadave** [comment]
If it's a big enough change and not related to "architecture" (software/infra/etc), I think the DR part of ADR is still important to capture. It's probably fine as an GH issue, but we might get some more consistency by capturing as an ADR. I don't want an over-bearing processes. Up to ya'll, just my 2c.

**Comment by jpetto** [comment]
I'm in favor of filing some sort of DR for any architecture/system/business logic change of significance. (I realize "of significance" is open to interpretation.) Personally, I find the "DR" part more important than that "A" part. 😉

**Comment by glogiotatidis** [comment]
Valid points folks. I will include an ADR.

**Comment by glogiotatidis** [comment]
Added ADR. Still need to work on the tests but code is there. Apologies for the huge PR folks!

**Comment by glogiotatidis** [comment]
Oh btw I removed summernote from this PR. Created #931 to track that

**Comment by glogiotatidis** [comment]
The migration script to convert ASRSnippets to the new system is in this gist
 
 https://gist.github.com/glogiotatidis/f143603e80e9044e0a9e67d777356147

This is a one-off, so not useful to commit in the repo. I'll get a detailed deployment plan  in #916

**Comment by metadave** [comment]
thanks for the ADR, I appreciate the context.

File: snippets/base/admin/adminmodels.py [hunk 2]
```diff
@@ -103,9 +104,337 @@ class AddonAdmin(admin.ModelAdmin):
...
+        # Don't allow image to be changed after set.
+        if obj and 'image' not in fields:
+            fields.append('image')
+
+        return fields
+
+    def snippets(self, obj):
+        """Snippets using this icon.
+
+        Needs that fancy code b/c icons have multiple relations with multiple Templates.
```
Comment (pmac) [inline snippets/base/admin/adminmodels.py]: Fancy! 🎩

Comment (metadave) [reply snippets/base/admin/adminmodels.py]: fancy, indeed

**Comment by glogiotatidis** [comment]
Updated the [migration script gist](https://gist.github.com/glogiotatidis/f143603e80e9044e0a9e67d777356147) to set default values to migrated snippets

**Comment by glogiotatidis** [comment]
The template are based on the ASR schemas,e.g.
e.g.
https://github.com/mozilla/activity-stream/blob/master/content-src/asrouter/templates/SimpleSnippet/SimpleSnippet.schema.json

File: snippets/base/admin/adminmodels.py (hunk 2 above)
Comment (glogiotatidis) [reply snippets/base/admin/adminmodels.py]: Moved the fancy code in `Icon` model to make it re-usable. ;)

**Comment by glogiotatidis** [comment]
OK this is done!

Added more tests, removed unused things and manually tested using a database dump from the real db.

@pmac final review at your time please! :)

File: snippets/urls.py [hunk 3]
```diff
@@ -43,7 +43,8 @@ def robots_txt(request):
     def serve_media(*args, **kwargs):
         response = static_serve(*args, **kwargs)
         response['Access-Control-Allow-Origin'] = '*'
-        if settings.BUNDLE_BROTLI_COMPRESS:
+        if ((settings.BUNDLE_BROTLI_COMPRESS and
```
[the remark this answers is not in the document]
Comment (glogiotatidis) [reply snippets/urls.py]: Intentional. One way to work around [E129](https://lintlyci.github.io/Flake8Rules/rules/E129.html) which works with my editor

**Comment by glogiotatidis** [comment]
> This all looks good to me Giorgos. It's a lot, and complex, so I could well have missed something, especially with real model inheritance... that's always tricky. It's a lot of good work though! As far as I'm concerned it's ready to give a good try +1

Agreed, I know it's complex. I've a detailed migration plan in #916 to make sure that I test all the things and I've communicated with the  LCM team that this is coming.

Thanks for your time and review!

**Pull request merged**